syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs

```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
 strBuilder quote = TokenStream.add =<< spanned parser where
     parser = do
         let isQuote      = (== quote)
-            isBodyChar c = c == quote || isEolBeginChar c
+            isBodyChar c = not $ c == quote || isEolBeginChar c
 
         quoteLen <- Source.length <$> takeWhile1 isQuote
 
-        let mkChunk       = \p -> TokenStream.add =<< spanned p
-            chunkPlain    = Atom.StrPlain <$> takeWhile1 (not . isBodyChar)
-            chunkQuote    = bodyQuotes =<< takeWhile1 isQuote
-            bodyQuotes qs = Atom.StrPlain qs <$ failIf (Source.length qs == quoteLen)
-            chunk         = spanned $ choice [chunkPlain, chunkQuote] -- lineBreak
-            ending        = void $ takeWhile1 isQuote
-            end           = ending <|> (TokenStream.add =<< invalid (pure $ Invalid.Literal $ Invalid.String Invalid.NoClosingMark))
-            -- body          = Ast.Str . concat <$> flexBlock (many1' chunk)
-            chunks        = (<>) <$> many1' chunk <*> (concat <$> flexBlock (many1' chunk))
-            body          = Ast.Str <$> chunks
-        if quoteLen == 2 then pure (Ast.Str mempty) else body <* end
+        let chunkBody    = takeWhile1 isBodyChar
+            chunkQuote   = bodyQuotes =<< takeWhile1 isQuote
+            bodyQuotes a = a <$ failIf (quoteLen == Source.length a)
+            chunkPlain   = spanned $ Atom.StrPlain . mconcat <$> many1' (choice [chunkBody, chunkQuote])
+
+            quotes       = void $ takeWhile1 isQuote
+            end          = quotes <|> noEnd
+            chunks       = many1' chunkPlain <|> (mergeStrChunks <$> strBlock chunkPlain)
+            noEnd        = (TokenStream.add =<< invalid (pure $ Invalid.Literal $ Invalid.String Invalid.NoClosingMark))
+            body         = Ast.Str <$> chunks
+        if quoteLen == 2 then pure (Ast.Str mempty) else (body <* end) <|> (Ast.Str mempty <$ noEnd)
 
 
 --------------------
 -- === Layout === --
 --------------------
 
-flexBlock1 :: Lexer a -> Lexer (NonEmpty a)
+strBlock1 :: forall t. Lexer (Spanned (Atom.StrChunk t)) -> Lexer [Spanned (Atom.StrChunk t)]
```

Comment:
Shouldn't this name be more descriptive?

Reply:
It's a block parser for string literals. It's a rather standard name for parser-like utils. Leaving as is for now, but if there are propositions for better names, I'd love to revisit this topic.